### PR TITLE
Upgrade Chalice to 1.24.1 (#3290)

### PR DESCRIPTION
**Pull request by jessebrennan**
https://github.com/DataBiosphere/azul/issues/3290

Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: src/azul/chalice.py [hunk 1]
```diff
@@ -65,6 +66,30 @@ def __init__(self, app_name, unit_test=False, spec=None):
         else:
             self._specs: Optional[MutableJSON] = None
         super().__init__(app_name, debug=config.debug > 0, configure_logs=False)
+        # The last handler is called first. We want logging to surround all
```
Comment (jessebrennan) [inline src/azul/chalice.py:69]: This comment is incorrect. I will fix it in the next cycle.

File: lambdas/layer/Makefile [hunk 2]
```diff
@@ -8,10 +8,6 @@ include ../lambdas.mk
 wheels: check_env
 	rm -rf vendor/jsonobject*
```
Comment (jessebrennan) [inline lambdas/layer/Makefile:9]: This needs to be changed to remove the whole contents of vendor or else add upgrading instructions to do that once so that we don't keep packaging the old wheels redundantly.

File: src/azul/chalice.py [hunk 3]
```diff
@@ -65,6 +66,30 @@ def __init__(self, app_name, unit_test=False, spec=None):
...
+        # other handler calls.
+        self.register_middleware(self.logging_middleware, 'http')
```
Comment (amarjandu) [inline src/azul/chalice.py:71]: IIRC the SQS -> Lambda interactions that the bundle notifications have use `sqs` as the event type rather than `http`, should we extend this logging middleware to have  `event_type=all` in order to record any type of event trigger that the lambdas? Not sure if we can do something like ('http', 'sqs') here.

**Review by amarjandu — CHANGES_REQUESTED** [review]
Looks good, just a question...

File: src/azul/chalice.py (hunk 3 above)
Comment (jessebrennan) [reply src/azul/chalice.py:71]: Glancing at the indexer lambdas, it seems like we already log when a notification comes in. Also middleware here would have to be changed since the SQS lambdas don't return a `response`.

If you think the SQS lambdas have insufficient logging, I think that would make sense as a separate issue.

**Review by amarjandu — APPROVED** [review]

File: src/azul/chalice.py (hunk 3 above)
Comment (jessebrennan) [reply src/azul/chalice.py:71]: @amarjandu, I'm confused about the eyes emoji, lol.

Comment (hannes-ucsc) [reply src/azul/chalice.py:71]: Me too.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: lambdas/layer/Makefile (hunk 2 above)
Comment (jessebrennan) [reply lambdas/layer/Makefile:9]: We should remove the whole contents of vendor because this caused a problem for @danielsotirhos where pyresist somehow ended up in there and caused duplicates.

File: src/azul/chalice.py [hunk 4]
```diff
@@ -65,6 +66,30 @@ def __init__(self, app_name, unit_test=False, spec=None):
...
+    def logging_middleware(self, event, get_response):
+        self._log_request()
+        response = get_response(event)
+        self._log_response(response)
+        return response
+
+    def authentication_middleware(self, event, get_response):
+        # We use the handler's private method to generate the response in case
+        # authentication fails
+        handler = RestAPIEventHandler(self.routes,
```
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/azul/chalice.py]: > The reason for switching to middleware is to reduce the hackyness of the current approach

This isn't the main reason for switching, but rather a hopeful side-effect. The reason we switched was because the method we were previously overriding in `Chalice()` no longer existed. Middleware seemed like the simplest way to get the same functionality.

> This is no less hacky than what we already do, it's less efficient and more surprising (two handler objects).

I agree, instantiating the handler again is hacky. The alternatives are:
- to duplicate the body of `_get_view_function_response()` (maybe not a bad idea) or
- subclass `RestAPIEventHander()` and change the behavior. 
I'll explain why below.

> I think the global error handler (_global_error_handler) is the outermost handler so the translation of any exception raised in _authenticate would still happen, even if we invoked it explicitly.

This is incorrect. `_global_error_hander()` simply packages the exception in a 500 response via `_unhandled_exception_to_response()`.

Anyway, assuming all of this stands, would you consider resetting the review counter?

[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/azul/chalice.py]: I'll bring this up in PL tomorrow.

[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/chalice.py]: > > The reason for switching to middleware is to reduce the hackyness of the current approach
> 
> This isn't the main reason for switching, but rather a hopeful side-effect. The reason we switched was because the method we were previously overriding in `Chalice()` no longer existed. Middleware seemed like the simplest way to get the same functionality.

Let me rephrase that. Authentication is a text book use case of middleware. We should be able to implement authentication using middleware without resorting to hacks.

> > I think the global error handler (_global_error_handler) is the outermost handler so the translation of any exception raised in _authenticate would still happen, even if we invoked it explicitly.
> 
> This is incorrect. `_global_error_hander()` simply packages the exception in a 500 response via `_unhandled_exception_to_response()`.

But the "packaging" of the exception in a 500 response is exactly the translation I was talking about. I don't see where my statement is incorrect. What are we not getting from the global handler that we're getting from your current approach?

> Anyway, assuming all of this stands, would you consider resetting the review counter?

I'll decide when it is clear that A) there is no better implementation or B) the implementation we end up using couldn't possibly have been implemented by you initially, because information wasn't available to you or I changed the specification mid-PR.

File: deployments/sandbox/environment.py [hunk 5]
```diff
@@ -59,8 +59,6 @@ def env() -> Mapping[str, Optional[str]]:
             ]
         }),
 
-        # FIXME: Add tooling to aid in prefix choice
-        #        https://github.com/DataBiosphere/azul/issues/3027
```
Comment (hannes-ucsc) [inline deployments/sandbox/environment.py:63]: COmmit needs ticket ref.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/chalice.py [hunk 6]
```diff
@@ -65,6 +66,24 @@ def __init__(self, app_name, unit_test=False, spec=None):
...
+        self._log_request()
+        response = get_response(event)
+        self._log_response(response)
+        return response
+
+    def authentication_middleware(self, event, get_response):
+        try:
+            self.__authenticate()
+        except ChaliceViewError as e:
+            response = Response(body={'Code': e.__class__.__name__, 'Message': str(e)},
```
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/azul/chalice.py:82]: I'll make this change. For my own edification, I'm curious as to why. I had copied the above code more or less verbatim from Chalice.

File: test/service/test_app_logging.py [hunk 7]
```diff
@@ -60,7 +60,9 @@ def test_request_logs(self):
                             level,
                             'Returning 200 response. To log headers and body, set AZUL_DEBUG to 1.'
                             if level == INFO else
-                            'Returning 200 response with headers {}. '
+                            'Returning 200 response with headers {"Access-Control-Allow-Origin": '
+                            '"*", "Access-Control-Allow-Headers": '
+                            '"Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key"}. '
```
[the remark this answers is not in the document]
Comment (jessebrennan) [reply test/service/test_app_logging.py:65]: @hannes-ucsc: "Jesse should investigate why headers weren't present before; timebox 30min."

[the remark this answers is not in the document]
Comment (jessebrennan) [reply test/service/test_app_logging.py:65]: Previously (before middleware), our insertion point for logging occurred by overriding the method called here: https://github.com/aws/chalice/blob/729b28543142636891ba01b2ac2e927cc38c3771/chalice/app.py#L1094

You can see, that the CORS headers were added soon after this method is called, so our logging statement was without them.

With middleware, the [_main_rest_api_handler()](https://github.com/aws/chalice/blob/ea779474e6dc193e21ea0a68c279e42634ad64b0/chalice/app.py#L1627) fully process the request before the middleware runs, thus the middleware is populated with all of the headers we expect.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
Operator, note drop! commit.

**Comment by hannes-ucsc** [comment]
Re: https://github.com/DataBiosphere/azul/pull/3323#discussion_r699761632

> > Anyway, assuming all of this stands, would you consider resetting the review counter?

> I'll decide when it is clear that A) there is no better implementation or B) the implementation we end up using couldn't possibly have been implemented by you initially, because information wasn't available to you or I changed the specification mid-PR.Re: 

I determine that neither A nor B apply, so review count stands.

**Comment by noah-aviel-dove** [comment]
@jessebrennan The commit title tagging is wrong. `[r]` is for reindex and `[R]` is for requirements updates. I believe the problem is perfunctory enough that I can fix it myself.

**Comment by noah-aviel-dove** [comment]
Conflicts resolved by operator
